File: exam_prep/IST603/assignment/part5/DivisionWithExceptions.py
"""
Write a Python program that:

- Accepts two numbers from the user.
- Divides the first number by the second.

Handle the following exceptions:
- ZeroDivisionError (division by zero)
- ValueError (invalid input when converting to a number)
- A general Exception block for any other unexpected errors.

Ensure a finally block is used to display a completion message.
"""


def divide_numbers() -> None:
    try:
        first = input("Enter the first number: ")
        second = input("Enter the second number: ")

        num1 = float(first)
        num2 = float(second)

        result = num1 / num2
        print(f"Result: {num1} / {num2} = {result}")
    except ZeroDivisionError:
        print("Error: Division by zero is not allowed.")
    except ValueError:
        print("Error: Invalid input. Please enter numeric values.")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
    finally:
        print("Division operation completed (Python).")


def main() -> None:
    divide_numbers()


if __name__ == "__main__":
    main()

File: exam_prep/IST603/assignment/part5/DivisionWithExceptions.java
package exam_prep.IST603.assignment.part5;

import java.util.Scanner;

/**
 * Write a Java program that:
 * <ul>
 *   <li>Accepts two numbers from the user.</li>
 *   <li>Divides the first number by the second.</li>
 * </ul>
 *
 * Handle the following exceptions:
 * <ul>
 *   <li>ArithmeticException (division by zero)</li>
 *   <li>NumberFormatException (invalid input)</li>
 *   <li>A general Exception block for unexpected errors</li>
 * </ul>
 *
 * Ensure a finally block is used to display a completion message.
 */
public class DivisionWithExceptions {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        try {
            System.out.print("Enter the first number: ");
            String firstInput = scanner.nextLine();

            System.out.print("Enter the second number: ");
            String secondInput = scanner.nextLine();

            int num1 = Integer.parseInt(firstInput);
            int num2 = Integer.parseInt(secondInput);

            int result = num1 / num2; // may throw ArithmeticException if num2 == 0

            System.out.println("Result: " + num1 + " / " + num2 + " = " + result);
        } catch (ArithmeticException e) {
            System.out.println("Error: Division by zero is not allowed.");
        } catch (NumberFormatException e) {
            System.out.println("Error: Invalid input. Please enter whole numbers.");
        } catch (Exception e) {
            System.out.println("An unexpected error occurred: " + e.getMessage());
        } finally {
            System.out.println("Division operation completed (Java).");
            scanner.close();
        }
    }
}

